### DISCUSSION TITLE
Using print dialog

### DISCUSSION THREAD

**Discussion by carlagomesah** [rubycdp/ferrum#343] [Q&A] [answered]
I'm with osx Monterey and chrome and I was able to trigger the printing dialog with:
```
browser.execute <<~JS
  window.print();
JS
```
after trying to trigger through keyboard combination, without success.

I tried to print to pdf with:
```
browser.window_opened_by do |window|
  window.use
  window.print_settings = {
    landscape: false,
    header_footer_enabled: false
  }
  window.print_to_pdf("output.pdf")
end
```
and although I've increased the timeout, that's still the error I receive.

Any thought on how to use the printing dialog windows?

**Comment by route** [comment]
What is the main goal you are trying to achieve? You don't write code to write code right :)? You want to save page as pdf?

**Reply by carlagomesah** [reply]
Yes, I want the pdf version of the page.
I can get something like that by using `pdf = browser.pdf(path: "something"`) but that's basically a screenshot. When I go through the printing dialog window and choose 'save as pdf' as printer, the page is quite different and that's the aim ( I know that further on it also asks the place where to save it but I don't even get to there).

I also changes my default printer to one that will `save as pdf`, added 👇  to try to click print after printing window opens :
```
browser.dialog.use do
  browser.click_button("Print")
end
```
but somehow I can't interact with the window.


```
require 'ferrum'

browser = Ferrum::Browser.new(headless: false, slowmo: (1), timeout: (60))
browser.goto('https://www.google.com')
browser.window_size

browser.execute <<~JS
  window.print();
JS

sleep(10)

browser.dialog.use do
  browser.click_button("Print")
end

sleep(50)

browser.quit
```

**Comment by route** [comment]
That's not a screenshot as it has lots of options https://github.com/rubycdp/ferrum#pdfoptions--string--boolean check native options as well, printBackground etc.

**Reply by carlagomesah** [reply]
I know it's not a screenshot, it has several options and in the end it's a pdf but the website behaves differently when using that _vs_ `cdm+p` as it shows a different output when generating the pdf. That's why I'm aiming to control the printing window :)

I've notice also that for this website, if I use `browser.pdf` it opens a blank window and the output is, well, blank. So I know this might be a special case and I'm not pointing fingers to ferrum. I'm just asking for ideas if anyone had the same problem // had the same aim 'cause I can't seem to have control of the printing window for any website.

Oh and in case it helps, the script opens 2 tabs, it will be running mostly on tab 2 and that's the one I intend to print to pdf.

**Reply by ttilberg** [reply]
Noting that [Puppeteer supports emulating print media](https://pptr.dev/api/puppeteer.page.emulatemediatype), I’m sure it can be done in Ferrum as well, we just need to figure out what it’s calling behind the scenes. 

this would probably allow you to set the page up with the printing CSS enabled, and then do a screenshot or pdf.

**Reply by ttilberg** [reply]
The CDP command appears to be https://chromedevtools.github.io/devtools-protocol/tot/Emulation/#method-setEmulatedMedia

I suspect the usage would look something like `browser.command('Emulation.setEmulatedMedia', 'print')`

I’m making a best guess and can’t check, as I’m away from a computer. But look in that direction!

**Reply by carlagomesah** [reply]
Thank you, will do it once I get to the computer also :)

**Reply by route** [reply]
I was just thinking that print dialog is not controllable through the JS or CDP and it's better to point you in `.pdf` direction rather than work on a hack solution.

**Reply by carlagomesah** [reply, marked as answer]
@route Ok, so maybe that's why I'm not able to make this work. In the end, for this particular case, after some compromise to accept what I was able to get, I'm using ` browser.pdf`
